spirv_msl.cpp

```diff
@@ -6792,6 +6804,25 @@ void CompilerMSL::emit_specialization_constants_and_structs()
 	// these types for purpose of iterating over them in ir.ids_for_type and friends.
 	auto loop_lock = ir.create_loop_soft_lock();
 
+	// Physical storage buffer pointers can have cyclical references,
+	// so emit forward declarations of them before other structs.
+	// Ignore type_id because we want the underlying struct type from the pointer.
+	ir.for_each_typed_id<SPIRType>([&](uint32_t /* type_id */, const SPIRType &type) {
+		if (type.basetype == SPIRType::Struct &&
+			type.pointer && type.storage == StorageClassPhysicalStorageBuffer &&
+			declared_structs.count(type.self) == 0)
+		{
+			statement("struct ", to_name(type.self), ";");
```

Comment:
This is also probably a good time to check for required Metal version for BDA.

Reply:
Although CPU access to device buffer memory addresses has just been added with the Metal 3 API, and MoltenVK will include a separate test for BDA for that, within SPIRV-Cross, the basics of pointers and `ulong` values in shader buffers are already separately covered by checks in SPIRV-Cross for MSL 2 & MSL 2.3 respectively, so I don't think we need to add anything specifically for BDA as a distinct feature.

Reply:
ok